### DISCUSSION TITLE
WSL2 Podman Install across the enterprise

### DISCUSSION THREAD

**Discussion by bhunsicker** [containers/podman#14899] [Q&A] [answered]
Hi Everyone,

I have an enterprise environment where I am not able to reach the Fedora repos or mirrors from a normal developer system.  I can install Podman on a special system and run the podman machine init and all the updates happen.  Then I need to distribute the image to my developers.

Currently my process is thus:
pull down the Fedora image
Expand it and update resolve.conf and add wsl.conf files
Add in company certificates and register them
repackage the system
Copy to special system (not blocked) and do a podman machine init using the repackaged tar.xz
stop the podman machine
do a WSL2 export

Copy the export and the appropriate config folders (<USER>\.local\share\containers, <USER>.config\share\containers, <USER>\appdata\roaming\containers\, <USER>\.ssh) to the new system
Install all the binaries into C:\Program Files\RedHat\Podman
Update the Path
copy the config folders to the users home
import the WSL2 export

At this point on the new system, I can run a podman machine list and see the machine.  I can start wsl -d podman-machine-default and do things in there.  I can even do a podman machine start.  However even though the response is that the machine is started, a podman machine inspect returns JSON with a stopped state for the machine, and a podman image returns this:

Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman socket: Get "http://d/v4.1.1/libpod/_ping": dial unix ///run/podman/podman.sock: connect: An invalid argument was supplied.

doing a podman system connection list returns the same values on both the old and new machine.

What am I missing?  

I cant use the podman machine init on the new machine as it is behind the firewall.

**Comment by baude** [comment]
@n1hility what say you?

**Comment by n1hility** [comment]
I suspect you are missing the AppData/Roaming/containers/containers.conf which contains the connection configurations. Let me know how that goes.

We are looking to add support for air gapped install soon.

**Comment by bhunsicker** [comment]
The connection configs are there.  They are one of the 4 folders I pulled across.

```
<USERHOME>\.local\share\containers
<USERHOME>\.config\share\containers
<USERHOME>\\appdata\roaming\containers
<USERHOME>\.ssh

```
```
> podman system connection list
Name                         URI                                                          Identity                                      Default
podman-machine-default       ssh://user@localhost:52669/run/user/1000/podman/podman.sock  C:\Users\bih8850\.ssh\podman-machine-default  true
podman-machine-default-root  ssh://root@localhost:52669/run/podman/podman.sock            C:\Users\bih8850\.ssh\podman-machine-default  false
```

**Comment by n1hility** [comment]
@bhunsicker Ah yes I see that now

- When you do a podman machine start, do you get any errors?
- Does podman machine ssh work, after start completes?
- If yes to the above, after you ssh in what does the output of systemctl status look like

**Comment by bhunsicker** [comment]
```
PS C:\codebase\wsl\wsl2images> podman machine start
Starting machine "podman-machine-default"

This machine is currently configured in rootless mode. If your containers
require root permissions (e.g. ports < 1024), or if you run into compatibility
issues with non-podman clients, you can switch using the following command:

        podman machine set --rootful

API forwarding for Docker API clients is not available due to the following startup failures.
        exec: "C:\\Program Files\\RedHat\\Podman\\win-sshproxy.exe": file does not exist

Podman clients are still able to connect.
Machine "podman-machine-default" started successfully
PS C:\codebase\wsl\wsl2images> podman machine ssh
Error: vm "podman-machine-default" is not running.
PS C:\codebase\wsl\wsl2images>
```

No errors that I recognize.  If I do an inspect I also see the status as stopped.

**Reply by n1hility** [reply]
Ok so that suggests that systemd bootstrap failed.

Can you paste the output of the following after a start:

- `wsl -l -v` 
- `wsl -d podman-machine-default id`
- `wsl -d podman-machine-default ps -ef`

Finally try

```
wsl -d podman-machine-default
# /root/bootstrap
```

if you see no specific failures after bootstrap, check the output of dmesg and see if anything pops up

**Comment by bhunsicker** [comment]
```
PS C:\codebase\wsl\wsl2images> podman machine stop
Error: "podman-machine-default" is not running
PS C:\codebase\wsl\wsl2images> wsl -l -v
  NAME                      STATE           VERSION
* podman-machine-default    Stopped         1
PS C:\codebase\wsl\wsl2images> podman machine start
Starting machine "podman-machine-default"

This machine is currently configured in rootless mode. If your containers
require root permissions (e.g. ports < 1024), or if you run into compatibility
issues with non-podman clients, you can switch using the following command:

        podman machine set --rootful

API forwarding for Docker API clients is not available due to the following startup failures.
        exec: "C:\\Program Files\\RedHat\\Podman\\win-sshproxy.exe": file does not exist

Podman clients are still able to connect.
Machine "podman-machine-default" started successfully
PS C:\codebase\wsl\wsl2images> wsl -l -v
  NAME                      STATE           VERSION
* podman-machine-default    Stopped         1
PS C:\codebase\wsl\wsl2images> wsl -d podman-machine-default id
uid=0(root) gid=0(root) groups=0(root)
PS C:\codebase\wsl\wsl2images> wsl -d podman-machine-default ps -ef
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  5 14:57 ?        00:00:00 /init
root         9     1  1 14:57 tty1     00:00:00 /init
root        10     9  4 14:57 tty1     00:00:00 ps -ef
PS C:\codebase\wsl\wsl2images> wsl -l -v
  NAME                      STATE           VERSION
* podman-machine-default    Running         1
PS C:\codebase\wsl\wsl2images> wsl -d podman-machine-default
[root@COMPUTER wsl2images]# /root/bootstrap
[root@SCOMPUTER wsl2images]#
```

**Reply by bhunsicker** [reply]
```
# dmesg
[    0.041567]  Microsoft 4.4.0-19041.1237-Microsoft 4.4.35
[    0.251348] <3>init: (1) ERROR: ConfigInitializeCommon:665: Failed to mount /usr/lib/wsl/drive
[    0.251355] : 19
[    0.251725] <3>init: (1) ERROR: ConfigInitializeCommon:665: Failed to mount /usr/lib/wsl/lib
[    0.251731] 19
[    0.573558] init: (1) WARNING: /etc/resolv.conf updating disabled in /etc/wsl.conf
[    0.614924] init: (1) WARNING: /etc/resolv.conf updating disabled in /etc/wsl.conf
```

**Comment by bhunsicker** [comment]
ok, so it looks like the WSL import of the image does not re-establish some of the mount points.

Got any advice on that one?

**Reply by bhunsicker** [reply]
And thank you for the suggestions thus far.  I would have been lost on that :)

**Reply by n1hility** [reply]
your welcome. OK so it looks like the issue is that this is being imported as WSL 1. Podman requires WSL2, can you try the following and then try running start again:

`wsl --set-version podman-machine-default 2`

**Reply by bhunsicker** [reply]
Seems the system I was given to test on did not have the WSL2 update either.  Loaded that and did the command you sent.  that fixed the problem.

MANY thanks :)

**Comment by n1hility** [comment, marked as answer]
You're welcome! 

Summarizing the answer for others:

Podman for Windows requires the usage of WSL2.

You can confirm the version of WSL by using the command 

`wsl -l -v`

If this does not report 2, you can use the command:

`wsl --set-version podman-machine-default 2`

Future versions of podman will detect this problem.
